components/provider/src/data_key.rs

```diff
@@ -185,29 +147,30 @@ impl From<&DataKey> for DataKeyComponents {
 
 #[test]
 fn test_to_string() {
+    use tinystr::tinystr4;
     struct TestCase {
         pub data_key: DataKey,
         pub expected: &'static str,
     }
     let cases = [
         TestCase {
-            data_key: icu_data_key!(plurals: cardinal@1),
+            data_key: data_key!(plurals, "cardinal", 1),
             expected: "plurals/cardinal@1",
         },
         TestCase {
-            data_key: icu_data_key!(x-private: cardinal@1),
-            expected: "x-private/cardinal@1",
-        },
-        TestCase {
-            data_key: icu_data_key!(x-maxlengthprivate: cardinal@1),
-            expected: "x-maxlengthprivate/cardinal@1",
+            data_key: DataKey {
+                category: DataCategory::PrivateUse(tinystr4!("priv")),
```

Comment:
can you add a factory method to generate it out of str instead?

Reply:
I could, but private use is not a common use case (we don't want people accidentally converting a string to private use), and you would lose the constness that we get by spelling it out with the macro.

Reply:
Yeah, I wouldn't do it as a default, but something like `DataCategory::newPrivateUse(&str)` or something like that. It's optional, and for later, just that using TinyStr in public API is a biit iffy for me.

Reply:
1. It's an enum, so the TinyStr field is public
2. I don't see anything wrong with having TinyStr be public
3. `DataCategory::newPrivateUse(&str)` can't be const, but the common use for DataKey and DataCategory is that they are const

Reply:
sgtm